pkg/nodeconfig/nodeconfig.go

```diff
@@ -288,6 +304,106 @@ func (nc *nodeConfig) Configure() error {
+	if err := nc.createKubeletConf(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createFilesFromIgnition creates any files it can from ignition: kubelet CA cert, cloud-config file
+func (nc *nodeConfig) createFilesFromIgnition() error {
+	ign, err := ignition.New(nc.client)
+	if err != nil {
+		return err
+	}
+	kubeletArgs, err := ign.GetKubeletArgs()
+	if err != nil {
+		return err
+	}
+
+	filesToTransfer := map[string]struct{}{
+		ignition.KubeletCACertPath: {},
+	}
+	if _, ok := kubeletArgs[ignition.CloudConfigOption]; ok {
+		filesToTransfer[ignition.CloudConfigPath] = struct{}{}
+	}
+	// For each new file in the ignition file check if is a file we are interested in and, if so, decode its contents
+	for _, ignFile := range ign.GetFiles() {
+		if _, ok := filesToTransfer[ignFile.Node.Path]; ok {
+			if ignFile.Contents.Source == nil {
+				return errors.Errorf("could not process %s: File is empty", ignFile.Node.Path)
+			}
+			contents, err := dataurl.DecodeString(*ignFile.Contents.Source)
+			if err != nil {
+				return errors.Wrapf(err, "could not decode %s", ignFile.Node.Path)
+			}
+			// Write to file on the Windows instance
+			fileName := filepath.Base(ignFile.Node.Path)
+			if err := nc.Windows.EnsureFileContent(contents.Data, fileName, windows.K8sDir); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// createBootstrapKubeconfig creates a kubeconfig for kubelet to initially communicate with the API server
+func (nc *nodeConfig) createBootstrapKubeconfig() error {
+	bootstrapSecret, err := nc.k8sclientset.CoreV1().Secrets(mcoNamespace).Get(context.TODO(), mcoBootstrapSecret,
+		meta.GetOptions{})
+	if err != nil {
+		return err
+	}
+	kubeconfig, err := generateKubeconfig(bootstrapSecret, nodeConfigCache.apiServerEndpoint)
+	if err != nil {
+		return err
+	}
+	kubeconfigData, err := json.Marshal(kubeconfig)
+	if err != nil {
+		return err
+	}
+	return nc.write(kubeconfigData, windows.BootstrapKubeconfigPath)
+}
+
+// createKubeletConf creates config file for kubelet, with Windows specific configuration.
+func (nc *nodeConfig) createKubeletConf() error {
+	clusterDNS, err := cluster.GetDNS(nc.clusterServiceCIDR)
+	if err != nil {
+		return err
+	}
+	kubeletConfig := generateKubeletConfiguration(clusterDNS)
+	kubeletConfigData, err := json.Marshal(kubeletConfig)
+	if err != nil {
+		return err
+	}
+	// Replace last character ('}') with comma
+	kubeletConfigData[len(kubeletConfigData)-1] = ','
+	// Appending this option is needed here instead of in the kubelet configuration object. Otherwise, when marshalling,
+	// the empty value will be omitted, so it would end up being incorrectly populated at service start time.
+	// Can be moved to kubelet configuration object with https://issues.redhat.com/browse/WINC-926
+	enforceNodeAllocatable := []byte("\"enforceNodeAllocatable\":[]}")
+	kubeletConfigData = append(kubeletConfigData, enforceNodeAllocatable...)
+
+	return nc.write(kubeletConfigData, windows.KubeletConfigPath)
```

Comment:
I'm thinking of returning the `kubeletConfigData` here and leaving the responsibility of writing the file to the `createBootstrapFiles` function. It may also help to unit test this function.

Reply:
I explored this option and decided against it. I'd like all the functions called from `createBootstrapFiles` to do the same thing -- if `createFilesFromIgnition` is writing data to a file, then `createBootstrapKubeconfig` and `createKubeletConf` should as well.

Reply:
We are adding a new component. If you find an existing pattern that does not look good, please change it though changing `createBootstrapKubeconfig` and `createKubeletConf` should be done in a separate PR or a separate commit at least.

Reply:
I weighed the two patterns and decided which was better...
However, if the team thinks `leaving the responsibility of writing the file to the createBootstrapFiles function` than the approach here, please take a look at the code in the next commit (0e2799bc0ce7bd2f7386d2d36779f4bb1eb2c1d4) and let me know.

Reply:
LGTM